ext/nnstreamer/tensor_filter/tensor_filter_ncnn.cc

```diff
@@ -0,0 +1,504 @@
+ncnn_subplugin::configure_instance (const GstTensorFilterProperties *prop)
+{
+  // get input / output info from properties
+  gst_tensors_info_copy (std::addressof (inputInfo), std::addressof (prop->input_meta));
+  gst_tensors_info_copy (std::addressof (outputInfo), std::addressof (prop->output_meta));
+
+  // check number of model files
+  if (prop->num_models > 2 || prop->num_models <= 0) {
+    throw std::invalid_argument (std::string ("Number of model files must be 1 or 2;")
+                                 + " Multiple model is not supported.");
+  }
+
+  // try to parse custom properties of the ncnn_subplugin
+  try {
+    // parse custom properties
+    parseCustomProperties (prop);
+  } catch (const std::invalid_argument &e) {
+    throw std::invalid_argument (
+        "Failed to parse custom property : " + std::string (e.what ()));
+  }
+
+  // decide use vulkan acceleration
+  if (std::find (prop->hw_list, prop->hw_list + prop->num_hw, ACCL_GPU)
+      != (prop->hw_list + prop->num_hw)) {
+    net.opt.use_vulkan_compute = true;
+    g_message ("accl = gpu\n");
+  } else {
+    net.opt.use_vulkan_compute = false;
+  }
+
+  // load model files
+  // ncnn returns nonzero value when an error occurs
+  if (prop->num_models == 1) {
+    if (net.load_param_bin (prop->model_files[0]))
+      throw std::invalid_argument (
+          "Failed to open the model file " + std::string (prop->model_files[0]));
+  } else {
+    if (net.load_param (prop->model_files[0]))
+      throw std::invalid_argument (
+          "Failed to open the param file " + std::string (prop->model_files[0]));
+    if (net.load_model (prop->model_files[1]))
+      throw std::invalid_argument (
+          "Failed to open the bin file " + std::string (prop->model_files[1]));
+  }
+
+  // get input layers from the ncnn network
+  const std::vector<int> &input_indexes = net.input_indexes ();
+  input_mats.clear ();
+  if (inputInfo.num_tensors != input_indexes.size ())
+    throw std::invalid_argument (
+        std::string ("Wrong number of input matrices")
+        + ": Found in argument = " + std::to_string (inputInfo.num_tensors)
+        + ", Found in model file = " + std::to_string (input_indexes.size ()));
+
+  // init input matrices
+  for (guint i = 0; i < inputInfo.num_tensors; i++) {
+    // get dimensions of the input matrix from inputInfo
+    const uint32_t *dim = inputInfo.info[i].dimension;
+    std::vector<int> shape;
+    while (*dim)
+      shape.push_back (*dim++);
+
+    // make ncnn matrix object
+    ncnn::Mat in;
+    switch (shape.size ()) {
+      case 1:
+        in = ncnn::Mat (shape[0]);
+        break;
+      case 2:
+        in = ncnn::Mat (shape[0], shape[1]);
+        break;
+      case 3:
+        in = ncnn::Mat (shape[0], shape[1], shape[2]);
+        break;
+      case 4:
+        in = ncnn::Mat (shape[0], shape[1], shape[2], shape[3]);
+        break;
+      default:
+        throw std::invalid_argument ("Wrong input dimension");
+    }
```

Comment:
@niley7464 : is it ok to support up to 4 ranks in a subplugin?

@chosanglyul : because nnstreamer latest supports up to 16 ranks, the error message might need to say "ncnn subplugin supports only up to 4 ranks and does not support input tensors of " + std::string (shape.size()) + "dimensions."

Reply:
https://github.com/nnstreamer/nnstreamer/blob/35642d7b1f060ce6304c0129813f88263715150f/ext/nnstreamer/tensor_filter/tensor_filter_nnfw.c#L469

nnfw subplugin uses it's own `NNFW_MAX_RANK` (6) so I think it's okay to support up to 4 rank.

I just realized that nnfw subplugin does not consider `GstTensorInfo *info` rank > `NNFW_MAX_RANK`.
Maybe this kind of exception handling has to be added in nnfw subplugin too.

Reply:
Updated to the suggested error message.